Content.Shared/Storage/EntitySystems/BinSystem.cs

```diff
@@ -0,0 +1,162 @@
+                return;
+            }
+        }
+    }
+
+    private void OnEntRemoved(EntityUid uid, BinComponent component, EntRemovedFromContainerMessage args)
+    {
+        component.Items.Remove(args.Entity);
+    }
+
+    private void OnInteractHand(EntityUid uid, BinComponent component, InteractHandEvent args)
+    {
+        if (args.Handled || !_timing.IsFirstTimePredicted)
+            return;
+
+        EntityUid? toGrab = component.Items.LastOrDefault();
+        if (!TryRemoveFromBin(uid, toGrab, component))
+            return;
+
+        _hands.TryPickupAnyHand(args.User, toGrab.Value);
+        _admin.Add(LogType.Pickup, LogImpact.Low,
+            $"{ToPrettyString(uid):player} removed {ToPrettyString(toGrab.Value)} from bin {ToPrettyString(uid)}.");
+        args.Handled = true;
+    }
+
+    private void OnAfterInteractUsing(EntityUid uid, BinComponent component, AfterInteractUsingEvent args)
+    {
+        if (args.Handled || !args.CanReach)
+            return;
+
+        if (!_timing.IsFirstTimePredicted)
+            return;
+
+        if (!TryInsertIntoBin(uid, args.Used, component))
+            return;
+
+        _admin.Add(LogType.Pickup, LogImpact.Low, $"{ToPrettyString(uid):player} inserted {ToPrettyString(args.User)} into bin {ToPrettyString(uid)}.");
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Inserts an entity at the top of the bin
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="toInsert"></param>
+    /// <param name="component"></param>
+    /// <returns></returns>
+    public bool TryInsertIntoBin(EntityUid uid, EntityUid toInsert, BinComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        if (component.Items.Count >= component.MaxItems)
+            return false;
+
+        if (component.Whitelist != null && !component.Whitelist.IsValid(toInsert))
+            return false;
+
+        component.ItemContainer.Insert(toInsert);
+        component.Items.Add(toInsert);
+        Dirty(component);
+        return true;
+    }
+
+    /// <summary>
+    /// Tries to remove an entity from the top of the bin.
+    /// </summary>
+    /// <param name="uid"></param>
+    /// <param name="toRemove"></param>
+    /// <param name="component"></param>
+    /// <returns></returns>
+    public bool TryRemoveFromBin(EntityUid uid, [NotNullWhen(true)] EntityUid? toRemove, BinComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        if (toRemove == null)
+            return false;
+
+        if (component.Items.Last() != toRemove)
+            return false;
+
+        if (!component.ItemContainer.Remove(toRemove.Value))
+            return false;
+
+        component.Items.Remove(toRemove.Value);
+        Dirty(component);
+        return true;
+    }
```

Comment:
Idk if this even compiles since im writing on github but i meant something like
```suggestion
    public bool TryRemoveFromBin(EntityUid uid, [NotNullWhen(true)] EntityUid? toRemove, BinComponent? component = null)
    {
        if (!Resolve(uid, ref component))
            return false;

        if (!(toRemove ??= component.Items.Last()))
            return false;

        if (!component.ItemContainer.Remove(toRemove.Value))
            return false;
		
        component.Items.Remove(toRemove.Value);
        Dirty(component);
        return _hands.TryPickupAnyHand(uid, toGrab.Value);;
    }
```